llama-index-core/llama_index/core/output_parsers/langchain.py

```diff
@@ -20,31 +20,30 @@ def __init__(
         """Init params."""
         self._output_parser = output_parser
         self._format_key = format_key
+        self._formatter = SafeFormatter()
 
     def parse(self, output: str) -> Any:
         """Parse, validate, and correct errors programmatically."""
-        # TODO: this object may be stringified by our upstream llmpredictor,
-        # figure out better
-        # ways to "convert" the object to a proper string format.
-        return self._output_parser.parse(output)
+        # Convert output to string if needed, then parse
+        output_str = str(output) if not isinstance(output, str) else output
+        return self._output_parser.parse(output_str)
 
     def format(self, query: str) -> str:
         """Format a query with structured output formatting instructions."""
         format_instructions = self._output_parser.get_format_instructions()
 
-        # TODO: this is a temporary hack. if there's curly brackets in the format
-        # instructions (and query is a string template), we need to
-        # escape the curly brackets in the format instructions to preserve the
-        # overall template.
-        query_tmpl_vars = {
-            v for _, v, _, _ in Formatter().parse(query) if v is not None
-        }
-        if len(query_tmpl_vars) > 0:
+        # Get template variables using SafeFormatter
+        query_tmpl_vars = self._formatter.parse(query)
+        
+        # If query contains template variables, escape curly braces in format_instructions
+        if query_tmpl_vars:
```

Comment:
We actually don't need to escape here, thats the benefit of using the safe formatter 👍🏻

Reply:
so you do not need the if? is it ok smt like this?
<img width="634" alt="Screenshot 2024-12-05 alle 20 36 51" src="https://github.com/user-attachments/assets/1b5d38b4-b367-4055-bc0f-4368dbee12e7">

Reply:
yea exactly, that looks great